incubator/validator-avro/src/main/java/io/aklivity/zilla/runtime/validator/avro/AvroValidator.java

```diff
@@ -14,132 +14,99 @@
+    protected final String subject;
+    protected final String format;
+    protected DatumReader<GenericRecord> reader;
+    protected DatumWriter<GenericRecord> writer;
+    protected final DirectBuffer valueRO = new UnsafeBuffer();
 
-    public AvroValidator(
+    private final Int2ObjectCache<Schema> cache;
+
+    protected AvroValidator(
         AvroValidatorConfig config,
-        ToLongFunction<String> resolveId,
         LongFunction<CatalogHandler> supplyCatalog)
     {
-        this.handlersById = new Long2ObjectHashMap<>();
         this.decoder = DecoderFactory.get();
-        this.catalogs = config.catalogs.stream().map(c ->
-        {
-            c.id = resolveId.applyAsLong(c.name);
-            handlersById.put(c.id, supplyCatalog.apply(c.id));
-            return c;
-        }).collect(Collectors.toList());
-        this.handler = handlersById.get(catalogs.get(0).id);
-        this.parser = new Schema.Parser();
-        this.catalog = catalogs.get(0).schemas.size() != 0 ? catalogs.get(0).schemas.get(0) : null;
-        this.subject = config.subject;
+        this.encoder = EncoderFactory.get();
+        CatalogedConfig cataloged = config.cataloged.get(0);
+        this.handler = supplyCatalog.apply(cataloged.id);
+        this.catalog = cataloged.schemas.size() != 0 ? cataloged.schemas.get(0) : null;
+        this.format = config.format;
+        this.subject = catalog != null && catalog.subject != null
+                ? catalog.subject
+                : config.subject;
+        this.cache = new Int2ObjectCache<>(1, 1024, i -> {});
     }
 
-    @Override
-    public boolean read(
-        DirectBuffer data,
-        int index,
-        int length)
+    protected Schema fetchSchema(
+        int schemaId)
     {
-        boolean status = false;
-        byte[] payloadBytes = new byte[length];
-        data.getBytes(0, payloadBytes);
-        ByteBuffer byteBuf = ByteBuffer.wrap(payloadBytes);
+        Schema schema = null;
 
-        if (byteBuf.get() == MAGIC_BYTE)
+        if (schemaId == 0)
         {
-            int schemaId = byteBuf.getInt();
-            int valLength = length - 1 - 4;
-            byte[] valBytes = new byte[valLength];
-            data.getBytes(length - valLength, valBytes);
-
-            String schema = handler.resolve(schemaId);
-
-            if (schema != null && validate(schema, valBytes))
-            {
-                status = true;
-            }
+            schemaId = handler.resolve(subject, catalog.version);
         }
-        return status;
-    }
 
-    @Override
-    public boolean write(
-        DirectBuffer data,
-        int index,
-        int length)
-    {
-        boolean status = false;
-        String schema = null;
-        int schemaId = catalog != null ? catalog.id : 0;
-
-        byte[] payloadBytes = new byte[length];
-        data.getBytes(0, payloadBytes);
-
-        if (schemaId > 0)
+        if (cache.containsKey(schemaId))
```

Comment:
Use `cache.computeIfAbsent(schemaId, this::resolveSchema)` and add a private method called `Schema resolveSchema(int schemaId)` that uses the `handler` to resolve the `schemaId` then parses and returns the `Schema`, or `null` if not resolved by `handler`.